### PR TITLE
feat(cli,plugin): did-you-run-the-tests verification nudge

### PR DESCRIPTION
**Pull request by alfhen**
## Summary

Feature 2 of the write-time nudge pair (Feature 1 = blast-radius nudge, #841, merged). Closes the honor-system gap in CLAUDE.md's "Verification Before Done" rule the same way `lien delta` and the blast-radius nudge already automate the complexity and `get_dependents` rules.

A session-scoped, append-only ledger (`<indexDir>/test-sessions/<sessionId>.jsonl`) records (a) edited files with associated tests and (b) Bash commands that looked like test runs. At session Stop, `lien verify-tests report` computes which edited files' associated tests were never observed running, and the Stop hook blocks once with an advisory `reason` — with an explicit escape hatch ("if you already ran them another way, disregard and stop again").

- `packages/cli/src/utils/test-run-matcher.ts` — pure `classifyTestCommand`/`computeUnverifiedFiles`: a conservative runner allow-list (npm/yarn/pnpm/bun/vitest/jest/mocha/pytest/go/cargo/rspec/phpunit/dotnet/deno/gradle/mvn + workspace-scoped forms + custom `test:*` script forms), broad-vs-scoped classification, exact-basename-or-stem coverage matching (see review update below). Any observed broad run silences the whole report.
- `packages/cli/src/utils/test-ledger.ts` — session ledger: append-only, sessionId validated against `^[A-Za-z0-9_-]+$`, `LIEN_TEST_VERIFY=off` kill switch (recording only — reads always work), cleaned up at SessionEnd + SessionStart's 24h GC alongside the existing `annotated-sessions/` cleanup.
- `packages/cli/src/cli/verify-tests-cmd.ts` — `lien verify-tests note-edit|note-run|report`. `note-edit` reuses `annotate-cmd.ts`'s `formatTestReminder` verbatim (refactored into a shared `scanTestAssociations` helper) so the reminder text is byte-identical to the old `annotate --tests-only` path it replaces.
- Hooks: `test-reminder.sh` rewired (calls `verify-tests note-edit` instead of `annotate --tests-only` — same reminder, now also records); new `test-run-note.sh` (`PostToolUse:Bash`, coarse shell pre-filter before ever shelling out to `lien`, silent recording only); new `test-verify-stop.sh` (`Stop`, `decision:block`+`reason`, `stop_hook_active` loop prevention **plus a ledger-based fallback**, see below).

## Empirical premise verification (mandatory)

**Premise: `lien verify-tests note-edit`'s reminder line is byte-identical to `lien annotate <file> --tests-only`.** Verified by building the CLI and diffing both commands against the same real file:

```
$ diff <(node packages/cli/dist/index.js annotate packages/cli/src/cli/annotate-cmd.ts --tests-only) \
       <(node packages/cli/dist/index.js verify-tests note-edit --session diff-test-session-2 --file packages/cli/src/cli/annotate-cmd.ts)
IDENTICAL (diff exit 0)
```
Both print: `Lien: you changed packages/cli/src/cli/annotate-cmd.ts — associated tests: packages/cli/src/cli/annotate-cmd.test.ts. Run them before completing.` This holds by construction (both call `formatTestReminder`), confirmed empirically as required.

## Dogfood evidence (real hook stdin shapes)

Global `lien` on PATH pointed at a different worktree's build (a concurrent session's own `npm link`), so I used a scoped PATH shim to this worktree's built CLI rather than touching the shared symlink — real stdin shapes piped directly into each script:

**(a) `test-run-note.sh` — real test command records, non-test command spawns no `lien` process:**
```
$ echo '{"session_id":"hook-dogfood-session","cwd":"...","hook_event_name":"PostToolUse","tool_name":"Bash","tool_input":{"command":"npm test -w @liendev/lien -- path/to/foo.test.ts"}}' | bash plugins/claude/hooks/test-run-note.sh
(stdout empty)
$ cat test-sessions/hook-dogfood-session.jsonl
{"kind":"run","timestamp":"...","command":"npm test -w @liendev/lien -- path/to/foo.test.ts"}
```
A `git status`/`ls -la` payload traced with `bash -x` never reaches the `lien` invocation line — the coarse shell `case` filter exits first.

**(b) `test-verify-stop.sh` — block+reason, silent-when-covered, stop_hook_active loop prevention:**
```
$ echo '{"session_id":"stop-hook-dogfood","cwd":"...","hook_event_name":"Stop","stop_hook_active":false}' | bash plugins/claude/hooks/test-verify-stop.sh
{"decision":"block","reason":"Before finishing: these files you edited this session have associated tests I\ndid not observe running in a Bash command:\n  • packages/cli/src/cli/annotate-cmd.ts → packages/cli/src/cli/annotate-cmd.test.ts\nIf you already ran them (watch mode, an IDE, or a wrapper this ledger can't see),\ndisregard and stop again. Otherwise, consider running them before you finish."}
```
After recording a covering scoped run (`vitest run packages/cli/src/cli/annotate-cmd.test.ts`), the same payload → empty stdout (silent, allows the stop). With `stop_hook_active:true` and the unverified edit still present → empty stdout (loop prevention).

**(c) Fail-open:** malformed (non-JSON) stdin and a payload missing `session_id` both exit 0 with no output, for both new hooks.

**Rewired `test-reminder.sh`:** a real `Edit` payload produced the identical `additionalContext` reminder **and** recorded the edit into the ledger in the same invocation — one process now does both.

## A/B #2 — run exactly as pre-registered (frozen protocol, honest result)

Protocol frozen in `.wip/ab-testverify/protocol.md` before any trial ran (per `.wip/nudge-design.md`'s pre-registration). N=8/arm, free CC subagents (`general-purpose`, `model: sonnet`, no isolation, no tools, no repo access), byte-identical prompt except the injected advisory block (captured verbatim by running the real CLI against a real indexed fixture repo, not hand-typed).

**Result: null — same ceiling-effect confound as the sibling blast-radius A/B (#841).** Both conditions: 8/8 ran at least one named test, 8/8 ran BOTH, 16/16 used a scoped (not broad) command. Three **control** trials — which received zero advisory text — independently cited CLAUDE.md's "Verification Before Done" section, and one (control-7) asserted *"the hook flagged these tests as unrun"* despite never being told a hook existed. Full write-up, raw per-trial results, and honest read: `docs/development/test-verification-nudge-ab.md`. No lift claim is made anywhere in the docs.

## Review round 1 — fixes applied (commit 7304513f)

1. **Fixed** `@liendev/cli` workspace reference (doesn't exist; real name is `@liendev/lien`) that was tripping the `docs-truth` CI linter — one plain correction in the architecture doc, one bracketed `[sic: ...]` correction preserving the verbatim A/B trial quote (raw untouched transcript stays in `.wip/ab-testverify/trials/results.md`). Re-ran `node .github/scripts/docs-truth.mjs` locally: `84 files checked, no violations.`
2. **Design deviation, declared:** `isCoveredByScope` replaced bidirectional substring matching with exact basename-or-stem equality. Substring containment let an unrelated run silently mark a file "covered" (`vitest run src/oauth.test.ts` covering an edit to `auth.ts`, because `"auth"` is a substring of `"oauth"`) — a worse failure mode than the false nag this feature is otherwise biased to avoid. Pinned with regression tests (`auth.ts`/`oauth.test.ts`, `user.ts`/`superuser.test.ts`, `a.ts`/`data.test.ts` — all now correctly nag) plus legit-coverage tests (exact test-file run, exact source-file token, cross-directory/`.spec` stem match). Documented as a dated deviation in both `docs/architecture/test-verification-nudge.md` and `.wip/nudge-design.md`.
3. **False-nag gaps fixed in `classifyTestCommand`**, all with new table-driven unit tests: (a) leading `VAR=value` env assignments (`CI=1 npm test`, `NODE_ENV=test vitest`) no longer defeat the anchored runner match; (b) `npm run test:*`/`yarn test:*`/`pnpm test:*` custom script forms recognized (covers this repo's own `npm run test:e2e:<lang>` convention); (c) `--config`/`-c` flag values and any bare `*.config.*` token are excluded from scope-token detection (`vitest --config vitest.config.ts` now correctly classifies `broad`).
4. **Fixed** the `test-run-note.sh` comment that stated the shell-filter/classifier superset relationship backwards — the shell filter is the (over-inclusive) superset; `classifyTestCommand` is the stricter subset. Reworded per the suggested fix.
5. **Robustness — `stop_hook_active` verified via a `claude-code-guide` agent, result: uncertain, not confirmed.** Three independent fetches of the official Stop-hook docs during this round returned three *different* field lists for the Stop event's stdin — one included `stop_hook_active`, two didn't (one instead listed an unrelated `stop_reason` field). Treated as unconfirmed rather than picking whichever result was convenient. **Decision: added a belt-and-braces fallback**, since the evidence didn't clear the "confirmed" bar the review asked for. `report` now records a `{kind:'blocked', timestamp}` ledger event the first time it emits a real advisory; `wasRecentlyBlocked` (pure, unit-tested) suppresses any block within a 10-minute window regardless of `stop_hook_active`. `stop_hook_active`'s own check is kept as the first line of defense (free if a given Claude Code version does populate it); the ledger-based window is the mechanism that actually holds if it doesn't. Dogfooded end-to-end: sent the identical `stop_hook_active:false` payload twice in a row (simulating "Claude Code never sends `stop_hook_active:true`") — first call blocked, second call was silent, confirming the fallback holds independent of that field ever arriving.
6. **Dismissed** the inline `logic_error` finding on `classifyTestCommand` (claimed `npm test -w @liendev/core -- src/foo.test.ts` misclassifies as broad) as a verified false positive — traced the loop and confirmed empirically for both the bot's cited example and a decisive variant (`.../src/sprocket.test.ts`, to rule out an accidental "foo" substring effect): both correctly produce `{"broad":false,"scopeTokens":["src/....test.ts"]}`. Reply posted on the inline comment with the repro; no code change.

Gate chain re-run after all fixes: `format:check` ✓ · `lint` ✓ (0 errors) · `typecheck` ✓ · `build` ✓ · `test` ✓ (parser 1077, core 378, review 1581, cli **1033** [+24 new], action 41, root 27 — all passing) · `node packages/cli/dist/index.js delta` ✓ exit 0 (advisory "worsened" notes only on the touched functions, all well under the cyclomatic/cognitive threshold of 15 — no new crossings). `docs-truth` linter: clean.

## Review round 2 — doc-only (commit f79b7464)

Re-verification passed on all fronts from round 1; one small doc-only addition. Added a "Known limitations" note to `docs/architecture/test-verification-nudge.md`: coverage matching is directory-blind (compares basenames/stems, not full paths), so two edited files sharing an associated-test basename — e.g. a per-module `index.ts` → `index.test.ts` convention repeated across packages — can have a scoped run of one silently cover the other too. Missed advisory, the safe direction this feature is biased toward throughout, and largely masked in practice by the any-broad-run silence rule (multi-file sessions typically end in a broad run). Names the future fix if it ever matters: fall back to directory-qualified path comparison when basenames collide among the pending files. Double-confirmed the round-1 `@liendev/cli` → `@liendev/lien` docs-drift fix is present at the pushed head (`git diff HEAD origin/... ` empty; `node .github/scripts/docs-truth.mjs` re-run fresh: `84 files checked, no violations`). `npx prettier --check` on the touched doc: clean. No code or test changes this round.

## Gates

`format:check` ✓ · `lint` ✓ (0 errors) · `typecheck` ✓ · `build` ✓ · `test` ✓ (all workspaces, 1033 cli tests) · `node packages/cli/dist/index.js delta` ✓ (no new crossings).

Durable docs: `docs/architecture/test-verification-nudge.md` (indexed in `docs/architecture/README.md`), `docs/development/test-verification-nudge-ab.md`. Also corrected `docs/architecture/claude-code-hook-channels.md`'s Stop-event section — a fresh check of the official hooks docs during the original build found Stop actually supports both `decision:block` (blocking) and `additionalContext` (non-blocking), contradicting the earlier "Stop ignores additionalContext" claim the frozen design was written on. `decision:block` is still the right choice for this feature's one-shot interruption; the doc now reflects the corrected channel behavior for future hook authors.

## Test plan

- [x] Unit tests: `test-run-matcher.test.ts` (60 tests — classification table incl. env-var stripping, `test:*` script forms, `--config` exclusion, `cd x && npm test` splitting, workspace-scoped forms, coverage decision incl. any-broad-run silence and the strict-matching regression repros), `test-ledger.test.ts` (27 tests — append-only, invalid session_id no-op, kill switch, `blocked` event round-trip), `verify-tests-cmd.test.ts` (29 tests — all three subcommand shapes, `wasRecentlyBlocked` pure-function table, block-suppression integration)
- [x] Real hook dogfood with actual stdin shapes, including the block-suppression fallback (see above)
- [x] A/B #2 run as frozen, raw rates reported honestly
- [x] Full gate chain green, `docs-truth` linter clean
- [x] One false positive dismissed with repro (inline reply), one wording bug fixed (inline reply)

---

### Lien Review

✅ **Good** - No complexity issues found.

> [!WARNING]
> **1 issue found** · Low Risk
>
> This PR adds a session-scoped did-you-run-the-tests verification nudge with a conservative classifier, append-only ledger, and fail-open hook integration. The core design is sound, but the environment-assignment stripping regex mishandles quoted values containing spaces, causing legitimate scoped test runs to be missed and potentially producing a false advisory at Stop.
>
> - Added `lien verify-tests note-edit|note-run|report` backed by a JSONL session ledger and pure `classifyTestCommand`/`computeUnverifiedFiles` classifier.
> - Rewired `test-reminder.sh` and added `test-run-note.sh`/`test-verify-stop.sh` hooks with belt-and-braces loop prevention.
> - One edge-case parsing gap: quoted env-variable assignments with spaces are not stripped before runner recognition.

⚠️ The documentation-truthfulness pass did not finish — it hit the token budget limit. Doc-claim verification is partial; code findings are unaffected.

<sup>Reviewed by [Lien Review](https://lien.dev) for commit f79b746. Updates automatically on new commits.</sup>

---

Attested: degraded:budget_starved · 0/1 comments posted · 387K/309K tokens

## Summary by CodeRabbit

- **New Features**
  - Added test-verification nudges that track edited files, observed test runs, and unverified tests during a session.
  - Added CLI commands to record edits, record test runs, and generate advisory reports.
  - Added optional stop notifications when associated tests have not been observed running.
  - Added a configuration switch to disable verification reminders.

- **Documentation**
  - Added architecture and development documentation for the verification workflow and hook behavior.

- **Tests**
  - Added comprehensive coverage for command handling, test detection, reporting, cleanup, and suppression behavior.

### CODE REVIEW

File: packages/cli/src/utils/test-run-matcher.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+/**
+ * Classify a raw Bash command string. `isTestRun` is false (broad/scopeTokens
+ * both empty/false) when no segment matches a recognized runner. When at
+ * least one matching segment carries no scoping argument, the whole
+ * classification is `broad` (a whole-suite run is present) even if another
+ * segment in the same command also named specific files.
+ */
+export function classifyTestCommand(command: string): TestRunClassification {
```
[the remark this answers is not in the document]
Comment (alfhen) [reply packages/cli/src/utils/test-run-matcher.ts:157]: Dismissing — verified false positive.

`extractPathLikeTokens`'s manual index loop only advances `i` past the workspace flag's OWN value (a single extra `i++` inside the `if (WORKSPACE_SCOPE_FLAGS.has(token) ...)` branch), not past the rest of the token stream — the `for` loop's own increment then continues scanning normally from the next token. Traced + confirmed empirically for the exact command cited here:

```
$ classifyTestCommand('npm test -w @liendev/core -- src/foo.test.ts')
{"isTestRun":true,"broad":false,"scopeTokens":["src/foo.test.ts"]}
```

and the decisive variant used to double-check the tokenizer isn't accidentally keying off the literal substring "foo":

```
$ classifyTestCommand('npm test -w @liendev/core -- src/sprocket.test.ts')
{"isTestRun":true,"broad":false,"scopeTokens":["src/sprocket.test.ts"]}
```

Both correctly classify as scoped (not broad) with the right scope token — the `--` separator is skipped as a generic dash-prefixed flag (its own `if (token.startsWith('-')) continue;` branch, which does NOT advance `i` an extra step), and `src/{foo,sprocket}.test.ts` is then evaluated and recognized as path-like. No change made to the matching loop.

File: plugins/claude/hooks/test-run-note.sh [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+# `lien` at all: a command matching none of these substrings can never be a
+# recognized test run (the precise classification in `classifyTestCommand`,
+# driven via `lien verify-tests note-run`, is a strict superset of what this
+# filter lets through), so a routine non-test Bash call (ls, git status,
+# cat, ...) spawns no `lien` process at all — the whole point of doing this
+# check in shell rather than always shelling out and letting the CLI decide.
+#
+# Emits NOTHING to the model on any path — recording only, never a warning.
+# Best-effort throughout; never fails the Bash call. Disable via
+# LIEN_TEST_VERIFY=off.
```
[the remark this answers is not in the document]
Comment (alfhen) [reply plugins/claude/hooks/test-run-note.sh:21]: Fixed — the superset relationship was stated backwards. Reworded the comment (matching the fix suggested here almost verbatim): the shell pre-filter is the strict superset of what `classifyTestCommand` ultimately recognizes, so every real test run the classifier would accept also passes the coarse filter (nothing is silently dropped), while the filter also lets through some non-test commands (like `echo test`) that the classifier's anchored patterns then correctly reject.
